src/index/sparse/sparse_index_node.cc

```diff
@@ -316,14 +327,127 @@ class SparseInvertedIndexNode : public IndexNode {
 
     sparse::BaseInvertedIndex<T>* index_{};
     std::shared_ptr<ThreadPool> search_pool_;
+    std::shared_ptr<ThreadPool> build_pool_;
 
     // if map_ is not nullptr, it means the index is mmapped from disk.
     char* map_ = nullptr;
     size_t map_size_ = 0;
 };  // class SparseInvertedIndexNode
 
+// Concurrent version of SparseInvertedIndexNode
+//
+// Thread safety: only the overridden methods are allowed to be called concurrently.
+template <typename T, bool use_wand>
+class SparseInvertedIndexNodeCC : public SparseInvertedIndexNode<T, use_wand> {
+ public:
+    explicit SparseInvertedIndexNodeCC(const int32_t& version, const Object& object)
+        : SparseInvertedIndexNode<T, use_wand>(version, object) {
+    }
+
+    Status
+    Add(const DataSetPtr dataset, std::shared_ptr<Config> cfg) override {
+        std::unique_lock<std::mutex> lock(mutex_);
```

Comment:
can `std::shared_mutex<>` from C++17 (https://en.cppreference.com/w/cpp/thread/shared_mutex) solve the problem of concurrent access for readers and writers in this particular use case?

Reply:
we want to avoid the starvation of writers. this impl guarantees that new read requests have to wait for already waiting write requests to finish.

Reply:
lgtm's, I'm assuming that you've checked that it won't deadlock